java-client/session/src/main/java/io/deephaven/client/impl/SessionImpl.java

```diff
@@ -69,156 +63,77 @@
-            if (!future.isDone()) {
-                future.completeExceptionally(
-                        new IllegalStateException("Observer completed without response"));
-            }
+    private static Optional<Duration> parseMillis(String x) {
+        try {
+            return Optional.of(Duration.ofMillis(Long.parseLong(x)));
+        } catch (NumberFormatException e) {
+            return Optional.empty();
         }
     }
 
     private final SessionImplConfig config;
-    private final SessionServiceStub sessionService;
-    private final ConsoleServiceStub consoleService;
-    private final ObjectServiceStub objectService;
-    private final InputTableServiceStub inputTableService;
-    private final ApplicationServiceStub applicationServiceStub;
-    private final Handler handler;
+    private final DeephavenChannel bearerChannel;
+    // Needed for downstream flight workarounds
+    private final BearerHandler bearerHandler;
     private final ExportTicketCreator exportTicketCreator;
     private final ExportStates states;
-    private volatile AuthenticationInfo auth;
     private final TableHandleManagerSerial serialManager;
     private final TableHandleManagerBatch batchManager;
+    private final ScheduledFuture<?> pingJob;
 
-    private SessionImpl(SessionImplConfig config, Handler handler, AuthenticationInfo auth) {
-        SessionCallCredentials credentials = new SessionCallCredentials();
-        this.auth = Objects.requireNonNull(auth);
-        this.handler = Objects.requireNonNull(handler);
+    private SessionImpl(SessionImplConfig config, DeephavenChannel bearerChannel, Duration pingFrequency,
+            BearerHandler bearerHandler) {
         this.config = Objects.requireNonNull(config);
-        this.sessionService = config.channel().session().withCallCredentials(credentials);
-        this.consoleService = config.channel().console().withCallCredentials(credentials);
-        this.objectService = config.channel().object().withCallCredentials(credentials);
-        this.inputTableService = config.channel().inputTable().withCallCredentials(credentials);
-        this.applicationServiceStub = config.channel().application().withCallCredentials(credentials);
+        this.bearerChannel = Objects.requireNonNull(bearerChannel);
+        this.bearerHandler = Objects.requireNonNull(bearerHandler);
         this.exportTicketCreator = new ExportTicketCreator();
-        this.states = new ExportStates(this, sessionService, config.channel().table().withCallCredentials(credentials),
-                exportTicketCreator);
+        this.states = new ExportStates(this, bearerChannel.session(), bearerChannel.table(), exportTicketCreator);
         this.serialManager = TableHandleManagerSerial.of(this);
         this.batchManager = TableHandleManagerBatch.of(this, config.mixinStacktrace());
+        this.pingJob = config.executor().scheduleAtFixedRate(
+                () -> bearerChannel.config().getConfigurationConstants(
```

Comment:
Cristian pointed out that when the server was unavailable that the client retried in an excessive repetitive manner. Have you verified that this no longer happens? I believed it would require setting alternate grpc call options.

Also, Colin pointed out that possibly ConfigService might run on another host (say, for a cluster of instances) - should we add a ping grpc call directly to the `SessionService` in the event that we're hitting envoy and being routed to different hosts?

Reply:
I've taken out that more "complicated" retry logic, and figured that using a fixed job that "pings" multiple times during the timeout interval is "good enough". (I don't see much value in some sort of backoff schedule... either we successfully reauth before the timeout interval, or we don't.)

I'd very much be in favor of a simple no-op "ping" RPC in `SessionService`; although, I think @niloc132 might have opinions? The general advice might be "use Handshake", but the code as written right now doesn't depend on Flight. Right now I'm essentially (ab)using `ConfigService#getConfigurationConstants` since I know the user is authorized for it.